app/net/nodemcu_mdns.c

```diff
@@ -0,0 +1,947 @@
+	memerr1:
+	/* free pbuf */
+	pbuf_free(p);
+	return;
+}
+
+/**
+ * close the UDP pcb .
+ */
+void ICACHE_FLASH_ATTR
+nodemcu_mdns_close(void)
+{
+	os_timer_disarm(&mdns_timer);
+
+	if (mdns_pcb != NULL) {
+		udp_remove(mdns_pcb);
+	}
+	if (mdns_payload) {
+	  os_free(mdns_payload);
+	}
+	mdns_payload = NULL;
+	mdns_pcb = NULL;
+	ms_info = NULL;
+}
+
+static void ICACHE_FLASH_ATTR
+mdns_set_host_description(const char *name)
+{
+	host_description = name;
+}
+
+static void ICACHE_FLASH_ATTR
+mdns_enable(void)
+{
+	if(mdns_flag == 0) {
+		udp_recv(mdns_pcb, mdns_recv, NULL);
+	}
+}
+
+static void ICACHE_FLASH_ATTR
+mdns_disable(void)
+{
+	if (mdns_flag == 1) {
+		udp_recv(mdns_pcb, NULL, NULL);
+	}
+}
+
+static void ICACHE_FLASH_ATTR
+mdns_set_hostname(char *name) {
+	if (name == NULL) {
+	  name = "nodemcu";
+	}
+	host_name = name;
+}
+
+static void ICACHE_FLASH_ATTR
+mdns_set_servername(const char *name) {
+	if (name == NULL) {
+		name = "nodemcu";
+	}
+	char tmpBuf[128];
+	os_sprintf(tmpBuf, "_%s._tcp.local", name);
+	if (server_name) {
+	  os_free(server_name);
+	}
+	server_name = c_strdup(tmpBuf);
+	service_name_with_suffix = server_name;
+}
+
+static void ICACHE_FLASH_ATTR
+mdns_server_unregister(void) {
+	struct ip_addr ap_host_addr;
+	struct ip_info ipconfig;
+	if(register_flag == 1){
+		if (igmp_leavegroup(&host_addr, &multicast_addr) != ERR_OK) {
+			MDNS_DBG("sta udp_leave_multigrup failed!\n");
+			return;
+		};
+		if(wifi_get_opmode() == 0x03 || wifi_get_opmode() == 0x02) {
+		   wifi_get_ip_info(SOFTAP_IF, &ipconfig);
+		   ap_host_addr.addr = ipconfig.ip.addr;
+		   if (igmp_leavegroup(&ap_host_addr, &multicast_addr) != ERR_OK) {
+				MDNS_DBG("ap udp_join_multigrup failed!\n");
+				return;
+			};
+		}
+		os_timer_disarm(&mdns_timer);
+		register_flag = 0;
+	}
+}
+
+static void ICACHE_FLASH_ATTR
+mdns_server_register(void) {
+
+	if (register_flag == 1) {
+		MDNS_DBG("mdns server is already registered !\n");
+		return;
+	} else if (igmp_joingroup(&host_addr, &multicast_addr) != ERR_OK) {
+		MDNS_DBG("udp_join_multigrup failed!\n");
+		return;
+	};
+	register_flag = 1;
+}
+
+static u8_t reg_counter;
+
+static void
+mdns_reg_handler_restart(void) {
+  reg_counter = 99;
+}
+
+static void ICACHE_FLASH_ATTR
+mdns_reg(struct nodemcu_mdns_info *info) {
+  mdns_send_service(info,0,0,0);
+  if (reg_counter++ > 10) {
+    mdns_send_service_type(0,0,0);
+    reg_counter = 0;
+  }
+}
+
+/**
+ * Initialize the resolver: set up the UDP pcb and configure the default server
+ * (NEW IP).
+ */
+void ICACHE_FLASH_ATTR
+nodemcu_mdns_init(struct nodemcu_mdns_info *info) {
+	/* initialize default DNS server address */
+	multicast_addr.addr = DNS_MULTICAST_ADDRESS;
+	struct ip_addr ap_host_addr;
+	struct ip_info ipconfig;
+	ms_info = info;		// Save the passed block. We need all the data forever
+
+	if (mdns_payload) {
+	  os_free(mdns_payload);
+	}
+	mdns_payload = (u8_t *) os_malloc(DNS_MSG_SIZE);
+	if (!mdns_payload) {
```

Comment:
This sort of logic produced extremely difficult-to-diagnose failures.  `nodemcu_init_info()`is a void function so  simply returning like this on allocation failure means that this init function simply fails silently.  No good.